repairDuplicateVersions.js

```diff
@@ -0,0 +1,447 @@
+        hostname: urlObj.hostname,
+        port: urlObj.port,
+        path: `${urlObj.pathname}${urlObj.search}`,
+        method,
+        agent: httpAgent,
+    }, res => {
+        if (method === 'HEAD') {
+            return cbOnce(null, res);
+        }
+        const chunks = [];
+        res.on('data', chunk => chunks.push(chunk));
+        res.once('end', () => {
+            const body = chunks.join('');
+            // eslint-disable-next-line no-param-reassign
+            res.body = body;
+            return cbOnce(null, res);
+        });
+        res.once('error', err => cbOnce(new Error(
+            'error reading response from HTTP request '
+                + `to ${url}: ${err.message}`
+        )));
+        return undefined;
+    });
+    req.once('error', err => cbOnce(new Error(
+        `error sending HTTP request to ${url}: ${err.message}`
+    )));
+    if (reqBody) {
+        req.setHeader('content-type', 'application/json');
+        req.setHeader('content-length', reqBody.length);
+        req.write(reqBody);
+    }
+    req.end();
+}
+
+function getSproxydAlias(cb) {
+    const url = `http://${SPROXYD_HOSTPORT}/.conf`;
+    httpRequest('GET', url, null, (err, res) => {
+        if (err) {
+            return cb(err);
+        }
+        if (res.statusCode !== 200) {
+            return cb(new Error(
+                `GET ${url} returned status ${res.statusCode}`
+            ));
+        }
+        const resp = JSON.parse(res.body);
+        sproxydAlias = resp['ring_driver:0'].alias;
+        return cb();
+    });
+}
+
+function readVerifyLog(cb) {
+    const logLines = readline.createInterface({ input: process.stdin });
+    logProgress('start reading verify log');
+    logLines.on('line', line => {
+        status.logLinesRead += 1;
+        try {
+            const parsedLine = JSON.parse(line);
+            if (parsedLine.message !== 'duplicate sproxyd key found') {
+                return undefined;
+            }
+            if (!parsedLine.objectUrl || !parsedLine.objectUrl2) {
+                log.error('malformed verify log line: missing fields', {
+                    lineNumber: status.logLinesRead,
+                });
+                return undefined;
+            }
+            objectsToRepair.push({
+                objectUrl: parsedLine.objectUrl,
+                objectUrl2: parsedLine.objectUrl2,
+            });
+        } catch (err) {
+            log.info('ignoring malformed JSON line');
+        }
+        return undefined;
+    });
+    logLines.on('close', () => {
+        logProgress('finished reading verify log');
+        cb();
+    });
+}
+
+function fetchObjectMetadata(objectUrl, cb) {
+    if (!objectUrl.startsWith('s3://')) {
+        return cb(new Error(`malformed object URL ${objectUrl}: must start with "s3://"`));
+    }
+    const bucketAndObject = objectUrl.slice(5);
+    const url = `http://${BUCKETD_HOSTPORT}/default/bucket/${bucketAndObject}`;
```

Comment:
Is there a chance that this script needs to be run at one of the sites that have enabled end-to-end TLS?

Reply:
Yes, it potentially could have to run at any of our S3 customers. Reading between the lines, I agree we should make it configurable with https as well.

Reply:
I opened an issue to keep track of it: https://scality.atlassian.net/browse/S3C-2807